ement-room.el

```diff
@@ -2817,151 +2809,113 @@ MSG is the message STRING was contained in."
-;;                                                            msg
-;;                                                            (lambda (a b)
-;;                                                              (string= (ement-event-id a)
-;;                                                                       (ement-event-id b)))
-;;                                                            :pred #'ement-event-p))
-;;                        (next? (cl-loop with w = (ewoc-next ement-ewoc node)
-;;                                        repeat num
-;;                                        if (or (not w)
-;;                                               (not (listp (ewoc-data w)))
-;;                                               (not (eq (car (ewoc-data w))
-;;                                                        'url-preview))
-;;                                               (not (equal (cadr (ewoc-data w))
-;;                                                           data)))
-;;                                        return t
-;;                                        do (setf w (ewoc-next ement-ewoc w))
-;;                                        finally return nil)))
-;;                   (when next?
-;;                     (with-silent-modifications
-;;                       (ewoc-enter-after ement-ewoc
-;;                                         node
-;;                                         `(url-preview ,data)))))))))
-
-(defun ement-room--insert-preview (buf msg related data num)
+(defun ement-room--get-preview (msg link num)
+  "Get preview of LINK. MSG is the message this link is from. NUM is the index for this link."
+  (let* ((id (ement-event-id msg))
+         (related-event (cl-loop with match = (make-ement-event :id id)
+                                 for e in (ement-room-timeline ement-room)
+                                 if (ement--events-equal-p match e)
+                                 return e))
+         (saved-buffer (current-buffer)))
+    (unless (or ement-room--preview-stop
+                (map-elt (ement-event-local related-event)
+                         'url-preview-n))
+      (ement-api ement-session "preview_url"
+        :endpoint-category "media" :params `((url ,link))
+        :then (lambda (data)
+                (when (buffer-live-p saved-buffer)
+                  (with-current-buffer saved-buffer
+                    (ement-room--insert-preview msg related-event num (ement-room--cleanse-url-preview data link)))))
+        :else (lambda (why) (ement-debug "Error in fetching preview: " why))))))
+
+(defun ement-room--cleanse-url-preview (data link)
+  "Ensure a url preview DATA always has a title and associated LINK."
+  (unless (alist-get 'og:url data) (push (cons 'og:url link) data))
+  (unless (alist-get 'og:title data)
+    (push (cons 'og:title (replace-regexp-in-string "\\<https?://\\|/.*$\\|/$" "" link)) data))
+  data)
+
+(defun ement-room--insert-preview (msg related num data)
```

Comment:
Again, it's probably not necessary to pass an index around.  This is Lisp, not C.  :)  As well, being async code, the data could have changed since the request was made.

Reply:
Hmm, how else would you suggest I do it? Because there can be multiple previews in the message, and it wouldn't know which one to insert if I don't pass it a number.

Reply:
I'd suggest using an alist keyed on the URL rather than indexing them with numbers.

Basically, I think the way it should work is this: when receiving a preview for an event, add its data to that event's `local` slot, then invalidate that event in the EWOC.  When the event is rendered in the EWOC, render any previews to strings and append them to the event's rendered string.

Reply:
I implemented it with alist keyed using URL at first, but then switched to keying by numbers is to provide information about the order. The first link in the message should have the first preview, the second should have the second, etc. Keying by URL alone would mean that previews would be randomly inserted depending on the order they were inserted into the alist, which would just be determined by the order they arrived from the network, which isn't great.

Reply:
I agree that the preview order should match the order of URLs in the message, but rather than use indexes, we can simply parse the URLs out of the message in-order when appending the previews.  This is something that isn't going to happen often at all, and it probably won't even take 1 ms for the CPU, so performance isn't a concern and we can keep the code as simple as possible.